utils/generator.py

```diff
@@ -0,0 +1,528 @@
+        """Return a dictionary with template data necessary to render an
+        interface a Python class.
+
+        This is used to render interfaces that are referenced by container
+        classes. The current list of rendered interfaces is passed as a second
+        argument to allow this method to add more interfaces to it as they are
+        discovered.
+
+        The returned format is as follows:
+
+        ```python
+        {
+            "name": "the class name to use for the interface class",
+            "parent": "the parent class name",
+            "args": [ # a Python description of each class attribute
+                "name": "the name of the attribute",
+                "type": "the Python type hint for the attribute",
+                "doc": ["formatted lines of documentation to add to class docstring"],
+                "required": bool,
+                "positional": bool,
+            ],
+        }
+        ```
+        """
+        type_ = schema.find_type(interface)
+        if type_["kind"] != "interface":
+            raise RuntimeError(f"Type {interface} is not an interface")
+        k = {"name": interface, "args": []}
+        while True:
+            for arg in type_["properties"]:
+                arg_type, _ = schema.get_attribute_data(arg)
+                k["args"].append(arg_type)
+            if "inherits" in type_ and "type" in type_["inherits"]:
+                if "parent" not in k:
+                    k["parent"] = type_["inherits"]["type"]["name"]
+                if type_["inherits"]["type"]["name"] not in interfaces:
+                    interfaces.append(type_["inherits"]["type"]["name"])
+                type_ = schema.find_type(
+                    type_["inherits"]["type"]["name"],
+                    type_["inherits"]["type"]["namespace"],
+                )
+            else:
+                break
+        return k
+
+
+def generate_query_py(schema, filename):
+    """Generate query.py with all the properties of `QueryContainer` as Python
+    classes.
+    """
+    classes = []
+    query_container = schema.find_type("QueryContainer", "_types.query_dsl")
+    for p in query_container["properties"]:
+        classes.append(schema.property_to_python_class(p))
+
+    with open(filename, "wt") as f:
+        f.write(query_py.render(classes=classes, parent="Query"))
+    print(f"Generated {filename}.")
+
+
+def generate_interfaces_py(schema, filename):
+    """Generate interfaces.py"""
+    classes = {}
+    schema.interfaces = sorted(schema.interfaces)
+    for interface in schema.interfaces:
+        if interface == "PipeSeparatedFlags":
+            continue  # handled as a special case
+        k = schema.interface_to_python_class(interface, schema.interfaces)
+        classes[k["name"]] = k
+
+    classes_list = []
+    for n in classes:
+        k = classes[n]
+        try:
+            classes_list.index(k)
+            continue
+        except ValueError:
+            pass
```

Comment:
Any reason not to use [LBYL](https://docs.python.org/3/glossary.html#term-LBYL) here?

```suggestion
        if k in classes_list:
            continue
```

Reply:
Yes, the logic in this function was more complex initially and I needed to know the index, but now I've got a much simpler solution so a simple check works. Thanks.